lib/spack/spack/solver/asp.py

```diff
@@ -1887,54 +1918,38 @@ def key_fn(item):
                 if deprecated:
                     self.deprecated_versions[pkg_name].add(v)
 
-            # All the preferred version from packages.yaml, versions in external
-            # specs will be computed later
-            version_preferences = packages_yaml.get(pkg_name, {}).get("version", [])
+            if pkg_name not in packages_yaml or "version" not in packages_yaml[pkg_name]:
+                continue
+
             version_defs = []
-            pkg_class = spack.repo.PATH.get_pkg_class(pkg_name)
-            for vstr in version_preferences:
+
+            for vstr in packages_yaml[pkg_name]["version"]:
                 v = vn.ver(vstr)
+
                 if isinstance(v, vn.GitVersion):
-                    version_defs.append(v)
+                    if not require_checksum or v.is_commit:
+                        version_defs.append(v)
                 else:
-                    satisfying_versions = self._check_for_defined_matching_versions(pkg_class, v)
-                    # Amongst all defined versions satisfying this specific
-                    # preference, the highest-numbered version is the
-                    # most-preferred: therefore sort satisfying versions
-                    # from greatest to least
-                    version_defs.extend(sorted(satisfying_versions, reverse=True))
+                    # TODO: stop re-defining existing versions, just relabel/adjust weights.
+                    matches = [x for x in self.possible_versions[pkg_name] if x.satisfies(v)]
+                    # TODO: use concretization order instead of version order
```

Comment:
Not clear what you mean by these comments.

Reply:
Let me remove the comment since it's irrelevant to the PR. I've explained it in a comment above: there's no need to duplicate versions if you can just mark/promote them. Also if a preference matches multiple existing versions, they should not be ordered by version, but by the original order `(preferred in package.py, not deprecated, not finite, version)`.